### ISSUE TITLE
ARM deployment with web app, continuous deployment and nodejs fails

### ISSUE THREAD

**Issue by morsh** [Azure/azure-xplat-cli#2618]
CLI Version: **0.9.X**
OS Type: Win
Installation via:  npm

Mode: **ARM**

Description:
`azure group deploy create` fails when ARM template has the following elements:

1. Web app
  1. appsettings [ type = config ]
  2. web [ type = sourcecontrol ]
  3. source control is git hub repository with node js code (I think you need to have 

Steps to reproduce:
1.  In the parameters file set the git hub repository to: https://github.com/morsh/simple-web-app.git
2.  Run `azure group deploy create -f template.json -e parameters resource-group-name deployment-name`

(I have added development modules from another project to reproduce the problem)

The template, parameters and deployment log (from the machine) files are attached:
[azuredeploy.parameters.zip](https://github.com/Azure/azure-xplat-cli/files/162866/azuredeploy.parameters.zip)

Deployment error:

info:    Executing command group deployment create
+ Initializing template configurations and parameters
+ Creating a deployment
info:    Created template deployment "deployment-1"
+ Waiting for deployment to complete
error:   Deployment provisioning state was not successful

**Comment by amarzavery** [comment]
@morsh - Did you try the group log show command.
Using it with --json switch will give you detailed error logs and help you find out the exact reason (which step in the deployment failed and for what reason).
```
D:\sdki>azure group log show -h
help:    Retrieves and shows logs for resource group operations
help:
help:    Usage: group log show [options] [name]
help:
help:    Options:
help:      -h, --help                        output usage information
help:      -v, --verbose                     use verbose output
help:      -vv                               more verbose with debug output
help:      --json                            use json output
help:      -n --name <name>                  the resource group name
help:      -a --all                          returns logs for all operations (including CRUD and deployment)
help:      -l --last-deployment              returns logs for the last deployment
help:      -d --deployment <name>            the name of the deployment whose logs you want to see
help:      -s --subscription <subscription>  the subscription identifier
help:
help:    Current Mode: arm (Azure Resource Management)
```

**Comment by morsh** [comment]
I can also see all this data in the azure portal.
This is the message I get:
`data:    Properties:           statusCode:    Conflict
                               statusMessage: {"status":"Failed","error":{"code
                               ":"
                               ResourceDeploymentFailure","message":"The
                               resource
                               operation completed with terminal provisioning
                               state '
                               Failed'."}}`
The real problem rises from the web app machine itself (This is the log I attached inside the zip file)

**Comment by morsh** [comment]
@amarzavery, Pinging... ;)

**Comment by amarzavery** [comment]
@ravbhatnagar @vivsriaus - Can you please take a look at this issue?

**Comment by vivsriaus** [comment]
@morsh Apologize for the delay here. Can you please run the 'azure group deployment create' command with -vv and include the captured requests/responses? I'm trying to understand if the issue here is the deployment failing or if the command is not returning the detailed errors.

**Comment by morsh** [comment]
[azure-deploy-failure.zip](https://github.com/Azure/azure-xplat-cli/files/195086/azure-deploy-failure.zip)
@vivsriaus thanks for the reply :)
I attached a zip with the following files:

- azure-cli-deployment-vv.log: the file you asked
- deployment failed web apps.png: screen shot of the deployment failue in azure
- template.json: the deployed template
- webapp-deployment-log.log: the deployment log taken from kudu of one of the failed machines

**Comment by vivsriaus** [comment]
@morsh Thanks for sending us the detailed logs. Our team is looking into the deployment failure and will respond shortly.
I noticed from the logs that you're using an older version of Azure CLI. We've made some changes recently to display detailed error messages when deployments fail, so I suggest upgrading to the latest version.

**Comment by morsh** [comment]
Hi @vivsriaus,
Pinging... :)

**Comment by vivsriaus** [comment]
@morsh Sorry I missed this. This is indeed a bug on our service end. We're working on a fix for this. I'll provide an update again early next week.

**Comment by morsh** [comment]
@vivsriaus. these are great news! Will be glad to know about updates.

**Comment by vivsriaus** [comment]
@morsh can you please verify that this is no longer an issue? I just got an update from our partner team indicating the issue was fixed yesterday.

**Comment by morsh** [comment]
Hi @vivsriaus,
Tried the deployment again, but got (sort of) new errors. All web apps I tried to deploy end up with 'Conflict' except for one that ends with 'Bad Request'.

I have an 'npm postinstall' script which used to run after every deployment. Now it doesn't run at all.

**azure cli response**:
error:   Deployment provisioning state was not successful
error:   The resource operation completed with terminal provisioning state 'Failed'.
error:   The resource operation completed with terminal provisioning state 'Failed'.
error:   The resource operation completed with terminal provisioning state 'Failed'.
error:   The resource operation completed with terminal provisioning state 'Failed'.
error:   The resource operation completed with terminal provisioning state 'Failed'.

**Portal result**:
genomics-scoring-staging2/web
Microsoft.Web/sites/sourcecontrols
Conflict
2016-04-17T09:02:53.0010623Z

genomics-trigger-staging2/web
Microsoft.Web/sites/sourcecontrols
Conflict
2016-04-17T09:02:43.1684736Z

genomics-docParser-staging2/web
Microsoft.Web/sites/sourcecontrols
Conflict
2016-04-17T09:02:50.3425668Z

genomics-graph-staging2/web
Microsoft.Web/sites/sourcecontrols
Conflict
2016-04-17T09:03:07.5559281Z

genomics-console-staging2/web
Microsoft.Web/sites/sourcecontrols
Conflict
2016-04-17T09:02:55.6639824Z

genomics-query-staging2/web
Microsoft.Web/sites/sourcecontrols
BadRequest
2016-04-17T08:42:46.6086949Z


I don't know if this is important, but i'm using azure internal (meaning there's an orange square at the top left corner labeled 'Preview UI') maybe there's a version difference?

If not, what else can we do to solve this?

**Comment by ravbhatnagar** [comment]
@morsh - Can you please confirm which version of CLI you have on your box? You should see similar errors between portal and CLI. As to debug why your deployment is failing, can you check the audit logs on the portal? That should give you some more details. If you can share the co-relation id for this deployment from the portal logs and I can help with the investigation.

**Comment by morsh** [comment]
This is the azure cli version:
0.9.14 (node: 0.12.7)

I still get this recurring problem on deployments, this are the same errors as before the update:

2016-04-17T08:45:48.0086216Z,npm ERR! network read ECONNRESET,,1
	2016-04-17T08:45:48.0086216Z,npm ERR! network This is most likely not a problem with npm itself,,1
	2016-04-17T08:45:48.0242450Z,npm ERR! network and is related to network connectivity.,,1
	2016-04-17T08:45:48.0398712Z,npm ERR! network In most cases you are behind a proxy or have bad network settings.,,1
	2016-04-17T08:45:48.0554953Z,npm ERR! network ,,1
	2016-04-17T08:45:48.0554953Z,npm ERR! network If you are behind a proxy&comma; please make sure that the,,1
	2016-04-17T08:45:48.0554953Z,npm ERR! network 'proxy' config is set properly.  See: 'npm help config',,1
	2016-04-17T08:45:48.0711198Z,npm ERR! Windows_NT 6.2.9200,,1
	2016-04-17T08:45:48.0867490Z,npm ERR! argv "D:\\Program Files (x86)\\nodejs\\4.2.3\\node.exe" "D:\\Program Files (x86)\\npm\\3.5.1\\node_modules\\npm\\bin\\npm-cli.js" "install" "--production",,1
	2016-04-17T08:45:48.0867490Z,npm ERR! node v4.2.3,,1
	2016-04-17T08:45:48.0867490Z,npm ERR! npm  v3.5.1,,1
	2016-04-17T08:45:48.1023700Z,npm ERR! code ECONNRESET,,1
	2016-04-17T08:45:48.1023700Z,npm ERR! errno ECONNRESET,,1
	2016-04-17T08:45:48.1179981Z,npm ERR! syscall read,,1
	

One of the app services does not have a deployment instance folder so you can't see the deployment logs. This is the one marked as "BadRequest".

**Comment by amarzavery** [comment]
@morsh - This seems to be an npm install problem..
Can you install the latest version of node.js in 4.x.x series and also install the latest version of azure-cli (0.9.20).

**Comment by morsh** [comment]
@amarzavery 
This is what I tried last time.
I will be back in two weeks and then I'll try that again and tell you if I succeeded.
Thanks!

**Comment by morsh** [comment]
@amarzavery 
I updated node.js to version 4.4.4 and azure to version 0.10.0.

I'm still getting the same errors. I've attached the deployment error file from one of the web apps.

[log (4).zip](https://github.com/Azure/azure-xplat-cli/files/256990/log.4.zip)

**Comment by morsh** [comment]
@amarzavery, @ravbhatnagar 
Pinging... :)

**Comment by amarzavery** [comment]
I looked at the error log provided in log(4).zip and it says callback called more than once.
```
Error: Callback called more than once.
npm verb stack     at errorHandler (c:\Program Files (x86)\npm\3.8.3\node_modules\npm\lib\utils\error-handler.js:110:16)
npm verb stack     at c:\Program Files (x86)\npm\3.8.3\node_modules\npm\lib\install.js:300:7
npm verb stack     at LOOP (c:\Program Files (x86)\npm\3.8.3\node_modules\npm\node_modules\slide\lib\chain.js:7:26)
npm verb stack     at c:\Program Files (x86)\npm\3.8.3\node_modules\npm\node_modules\slide\lib\chain.js:18:7
npm verb stack     at c:\Program Files (x86)\npm\3.8.3\node_modules\npm\lib\install.js:689:7
npm verb stack     at Function.ls.fromTree (c:\Program Files (x86)\npm\3.8.3\node_modules\npm\lib\ls.js:104:3)
npm verb stack     at c:\Program Files (x86)\npm\3.8.3\node_modules\npm\lib\install.js:687:8
npm verb stack     at c:\Program Files (x86)\npm\3.8.3\node_modules\npm\node_modules\iferr\index.js:13:50
npm verb stack     at c:\Program Files (x86)\npm\3.8.3\node_modules\npm\lib\install\deps.js:121:5
npm verb stack     at LOOP (c:\Program Files (x86)\npm\3.8.3\node_modules\npm\node_modules\slide\lib\chain.js:7:26)
npm verb cwd D:\home\site\wwwroot
npm ERR! Windows_NT 6.2.9200
npm ERR! argv "D:\\Program Files (x86)\\nodejs\\4.2.4\\node.exe" "D:\\Program Files (x86)\\npm\\3.8.3\\node_modules\\npm\\bin\\npm-cli.js" "install" "--verbose" "--production"
npm ERR! node v4.2.4
npm ERR! npm  v3.8.3

npm ERR! Callback called more than once.
npm ERR! 
npm ERR! If you need help&comma; you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm verb exit [ 1&comma; true ]

npm ERR! Please include the following file with any support request:
npm ERR!     D:\home\site\wwwroot\npm-debug.log
C:\Program Files (x86)\SiteExtensions\Kudu\54.50506.2209\bin\scripts\starter.cmd deploy.cmd,,2


```

**Comment by amarzavery** [comment]
@ravbhatnagar @vivsriaus - Can you take a look at this issue?

**Comment by morsh** [comment]
@amarzavery, @ravbhatnagar 
Still pinging... :)

**Comment by amarzavery** [comment]
@davidebbo - Can you take a look at the webapp deployment issue? Any clue/help would be great.

**Comment by davidebbo** [comment]
Your client side node/npm is clearly not relevant here.

Please try running the `npm install` from [Kudu Console](https://github.com/projectkudu/kudu/wiki/Kudu-console) in your d:\home\site\wwwroot folder, using the same node & npm versions (4.2.4 and 3.8.3). Do you get the same error?

This will help isolate from anything related to xplat CLI, ARM templates and git deployment.

**Comment by davidebbo** [comment]
Also, try upgrading to the latest Node available on Azure, which right now is `v4.4.1`. And npm `v3.8.6`.

**Comment by morsh** [comment]
Hi @davidebbo @amarzavery @ravbhatnagar,
A little update:
We can now deploy web apps and web jobs with continuous integration and the process works great.

One problem that remains - we have to perform two deployments:
1. First deployment: continuous integration in deployed but application settings are not
2. Second deployment (exactly the same command): application settings are deployed

**Comment by morsh** [comment]
Versions: npm - 4.4.4; azure-cli - 3.10.6

**Comment by morsh** [comment]
p.s. This might be a a duplicate of https://github.com/Azure/azure-xplat-cli/issues/2925 which has the same issue only with function apps (but the same scenario is concurrent with simple web apps)

**Comment by davidebbo** [comment]
@morsh I don't quite follow your scenario. Generally, continuous deployment (e.g. via GitHub or Bitbucket) never affects the Web App's  App Settings. Modifying App Settings requires separate calls the to different ARM APIs.

**Comment by morsh** [comment]
@davidebbo 
I mean having both resources in the same deployment template as follows:
[azuredeploy.json.txt](https://github.com/Azure/azure-xplat-cli/files/479741/azuredeploy.json.txt)

Notice that [Microsoft.Web/sites/webAppName] has two resources [config/appsettings] and [sourcecontrols/web]. 

the sourcecontrol is deployed in the first deployment while the appsettings is deployed only in the second deployment.

**Comment by morsh** [comment]
I mean exactly that.
The source control - yeah, that's the expected behavior.

**Comment by davidebbo** [comment]
This might be some race condition. Instead of doing the 3 actions in parallel, try adding dependencies and run things in this order:

1. web/sourcecontrol
2. config/appsettings
3. config/web

E.g. from 2 to 1 would look like:

    "[resourceId('Microsoft.Web/sites/sourcecontrols', variables('webAppName'), 'web')]"

**Comment by morsh** [comment]
Yes, that works :)
Thanks!

**Comment by davidebbo** [comment]
Great!
